### PR TITLE
[EOC-247] Refactor members reducer from array to object with

### PR DESCRIPTION
**Pull request by alemikolo**

### CODE REVIEW

File: client/src/modules/cohort/model/reducer.js [hunk 1]
```diff
@@ -3,57 +3,44 @@ import _keyBy from 'lodash/keyBy';
...
-const membersReducer = (state = [], action) => {
+const membersReducer = (state = {}, action) => {
   switch (action.type) {
     case CohortActionTypes.ADD_MEMBER_SUCCESS: {
       const {
         payload: { data }
       } = action;
 
-      return [...state, data];
+      return { [data._id]: data, ...state };
```
Comment (adamklepacz) [inline client/src/modules/cohort/model/reducer.js]: destructure _id from data

File: client/src/modules/list/model/reducer.js [hunk 2]
```diff
@@ -9,79 +9,64 @@ import {
...
-const membersReducer = (state = [], action) => {
+const membersReducer = (state = {}, action) => {
   switch (action.type) {
     case ListActionTypes.ADD_VIEWER_SUCCESS: {
       const {
         payload: { data }
       } = action;
 
-      return [...state, data];
+      return { [data._id]: data, ...state };
```
Comment (adamklepacz) [inline client/src/modules/list/model/reducer.js]: destructure _id

File: client/src/modules/cohort/model/selectors.js [hunk 3]
```diff
@@ -36,5 +36,5 @@ export const getArchivedCohorts = createSelector(
 
 export const getMembers = createSelector(
   getCohortDetails,
-  cohort => cohort && _keyBy(cohort.members, '_id')
+  cohort => cohort && cohort.members
```
Comment (marekrozmus) [inline client/src/modules/cohort/model/selectors.js]: this should be only: `cohort => cohort.members` if no members then empty object should be returned. Why checking if cohort exists?

File: client/src/modules/list/model/reducer.js [hunk 4]
```diff
@@ -9,79 +9,64 @@ import {
...
-          ? {
-              ...member,
-              isMember: false,
-              isOwner: false
-            }
-          : member
-      );
+      return {
+        ...state,
+        [userId]: { ...state[userId], isOwner: false, isMember: false }
```
Comment (marekrozmus) [inline client/src/modules/list/model/reducer.js]: if member role is removed then iwner role is also removed? What i n case if we are the only owner and we remove member role?

File: client/src/modules/list/model/selectors.js [hunk 5]
```diff
@@ -106,5 +106,5 @@ export const getCohortArchivedLists = createSelector(
 
 export const getMembers = createSelector(
   getList,
-  list => list && _keyBy(list.members, '_id')
+  list => list && list.members
```
Comment (marekrozmus) [inline client/src/modules/list/model/selectors.js]: same issue as with cohort

File: client/src/modules/list/model/reducer.js (hunk 4 above)
Comment (alemikolo) [reply client/src/modules/list/model/reducer.js]: We have to be a member if we want to be an owner. We can't have a higher level role without having a lower level one. We can't remove a member role when we are the only owner.

File: client/src/modules/cohort/model/selectors.js (hunk 3 above)
Comment (alemikolo) [reply client/src/modules/cohort/model/selectors.js]: without this check we get an error: Uncaught TypeError: Cannot read property 'members' of undefined

File: client/src/modules/list/model/selectors.js (hunk 5 above)
Comment (alemikolo) [reply client/src/modules/list/model/selectors.js]: without this check we get an error: Uncaught TypeError: Cannot read property 'members' of undefined

File: client/src/modules/list/model/reducer.js [hunk 6]
```diff
@@ -13,10 +13,13 @@ const membersReducer = (state = {}, action) => {
   switch (action.type) {
     case ListActionTypes.ADD_VIEWER_SUCCESS: {
       const {
-        payload: { data }
+        payload: {
+          data,
+          data: { _id: id }
```
Comment (adamklepacz) [inline client/src/modules/list/model/reducer.js]: just _id

File: client/src/modules/cohort/model/reducer.js [hunk 7]
```diff
@@ -7,10 +7,13 @@ const membersReducer = (state = {}, action) => {
   switch (action.type) {
     case CohortActionTypes.ADD_MEMBER_SUCCESS: {
       const {
-        payload: { data }
+        payload: {
+          data,
+          data: { _id: id }
```
Comment (adamklepacz) [inline client/src/modules/cohort/model/reducer.js]: just _id

File: client/src/modules/cohort/model/selectors.js (hunk 3 above)
Comment (marekrozmus) [reply client/src/modules/cohort/model/selectors.js]: ok, let's leave it as it is for now - we will need to review cohort component and it's model/selectors

**Pull request merged**